# GET

GET is one of the most common HTTP methods you’ll use when working with REST APIs. This method allows you to retrieve resources from a given API. GET is a read-only operation, so you shouldn’t use it to modify an existing resource.

To test out GET and the other methods in this section, you’ll use a service called JSONPlaceholder. This free service provides fake API endpoints that send back responses that requests can process.

To try this out, start up the Python REPL and run the following commands to send a GET request to a JSONPlaceholder endpoint:

In [1]:
import requests

In [2]:
api_url = "https://jsonplaceholder.typicode.com/todos/1"

In [3]:
response = requests.get(api_url)

In [4]:
response.json()

{'userId': 1, 'id': 1, 'title': 'delectus aut autem', 'completed': False}

This code calls requests.get() to send a GET request to /todos/1, which responds with the todo item with the ID 1. Then you can call .json() on the response object to view the data that came back from the API.

The response data is formatted as JSON, a key-value store similar to a Python dictionary. It’s a very popular data format and the de facto interchange format for most REST APIs.

In [5]:
response.status_code

200

In [6]:
response.headers["Content-Type"]

'application/json; charset=utf-8'

Here, you access response.status_code to see the HTTP status code. You can also view the response’s HTTP headers with response.headers. This dictionary contains metadata about the response, such as the Content-Type of the response.

# POST

Now, take a look at how you use requests to POST data to a REST API to create a new resource. You’ll use JSONPlaceholder again, but this time you’ll include JSON data in the request. Here’s the data that you’ll send:

In [7]:
api_url = "https://jsonplaceholder.typicode.com/todos"

In [8]:
todo = {"userId": 1, "title": "Buy milk", "completed": False}

In [9]:
response = requests.post(api_url, json=todo)

In [10]:
response.json()

{'userId': 1, 'title': 'Buy milk', 'completed': False, 'id': 201}

In [11]:
response.status_code

201

If you don’t use the json keyword argument to supply the JSON data, then you need to set Content-Type accordingly and serialize the JSON manually. Here’s an equivalent version to the previous code:

In [12]:
import requests
import json
api_url = "https://jsonplaceholder.typicode.com/todos"
todo = {"userId": 1, "title": "Buy milk", "completed": False}
headers =  {"Content-Type":"application/json"}
response = requests.post(api_url, data=json.dumps(todo), headers=headers)
response.json()

{'userId': 1, 'title': 'Buy milk', 'completed': False, 'id': 201}

In [13]:
response.status_code

201

Note: json.dumps() comes from the json package in the standard library. This package provides useful methods for working with JSON in Python.

# PUT

Beyond GET and POST, requests provides support for all the other HTTP methods you would use with a REST API. The following code sends a PUT request to update an existing todo with new data. Any data sent with a PUT request will completely replace the existing values of the todo.

You’ll use the same JSONPlaceholder endpoint you used with GET and POST, but this time you’ll append 10 to the end of the URL. This tells the REST API which todo you’d like to update:

In [14]:
import requests

In [15]:
api_url = "https://jsonplaceholder.typicode.com/todos/10"

In [16]:
response = requests.get(api_url)

In [17]:
response.json()

{'userId': 1,
 'id': 10,
 'title': 'illo est ratione doloremque quia maiores aut',
 'completed': True}

In [18]:
todo = {"userId": 2, "title": "Wash car", "completed": True}

In [19]:
response = requests.put(api_url, json=todo)

In [20]:
response.json()

{'userId': 2, 'title': 'Wash car', 'completed': True, 'id': 10}

In [21]:
response.status_code

200

Here, you first call requests.get() to view the contents of the existing todo. Next, you call requests.put() with new JSON data to replace the existing to-do’s values. You can see the new values when you call response.json(). Successful PUT requests will always return 200 instead of 201 because you aren’t creating a new resource but just updating an existing one.

### Status Codes

<table class="table table-hover">
<thead>
<tr>
<th>Code</th>
<th>Meaning</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>200</code></td>
<td>OK</td>
<td>The requested action was successful.</td>
</tr>
<tr>
<td><code>201</code></td>
<td>Created</td>
<td>A new resource was created.</td>
</tr>
<tr>
<td><code>202</code></td>
<td>Accepted</td>
<td>The request was received, but no modification has been made yet.</td>
</tr>
<tr>
<td><code>204</code></td>
<td>No Content</td>
<td>The request was successful, but the response has no content.</td>
</tr>
<tr>
<td><code>400</code></td>
<td>Bad Request</td>
<td>The request was malformed.</td>
</tr>
<tr>
<td><code>401</code></td>
<td>Unauthorized</td>
<td>The client is not authorized to perform the requested action.</td>
</tr>
<tr>
<td><code>404</code></td>
<td>Not Found</td>
<td>The requested resource was not found.</td>
</tr>
<tr>
<td><code>415</code></td>
<td>Unsupported Media Type</td>
<td>The request data format is not supported by the server.</td>
</tr>
<tr>
<td><code>422</code></td>
<td>Unprocessable Entity</td>
<td>The request data was properly formatted but contained invalid or missing data.</td>
</tr>
<tr>
<td><code>500</code></td>
<td>Internal Server Error</td>
<td>The server threw an error when processing the request.</td>
</tr>
</tbody>
</table>

# PATCH

Next up, you’ll use requests.patch() to modify the value of a specific field on an existing todo. PATCH differs from PUT in that it doesn’t completely replace the existing resource. It only modifies the values set in the JSON sent with the request.

You’ll use the same todo from the last example to try out requests.patch(). Here are the current values:

In [22]:
import requests

In [23]:
api_url = "https://jsonplaceholder.typicode.com/todos/10"

In [24]:
todo = {"title": "Mow lawn"}

In [25]:
response = requests.patch(api_url, json=todo)

In [26]:
response.json()

{'userId': 1, 'id': 10, 'title': 'Mow lawn', 'completed': True}

In [27]:
response.status_code

200

When you call response.json(), you can see that title was updated to Mow lawn.

# DELETE

Last but not least, if you want to completely remove a resource, then you use DELETE. Here’s the code to remove a todo:

In [28]:
import requests

In [29]:
api_url = "https://jsonplaceholder.typicode.com/todos/10"

In [30]:
response = requests.delete(api_url)

In [31]:
response.json()

{}

In [32]:
response.status_code

200

You call requests.delete() with an API URL that contains the ID for the todo you would like to remove. This sends a DELETE request to the REST API, which then removes the matching resource. After deleting the resource, the API sends back an empty JSON object indicating that the resource has been deleted.

The requests library is an awesome tool for working with REST APIs and an indispensable part of your Python tool belt. In the next section, you’ll change gears and consider what it takes to build a REST API.